File: Components/Strata.Foundation.Core/main/java/strata/foundation/core/value/PostalAddressBuilder.java
//////////////////////////////////////////////////////////////////////////////
// PostalAddressBuilder.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.value;

/**
 * <p>
 * A fluent builder for constructing {@link PostalAddress} instances one
 * field at a time. Each setter returns {@code this} so calls can be
 * chained, and any field left unset is passed through as {@code null},
 * which {@link PostalAddress} then normalizes to an empty string.
 * </p>
 * <p>
 * <h4>Examples</h4>
 * <pre>
 * PostalAddress address =
 *     new PostalAddressBuilder()
 *         .setAddress("123")
 *         .setStreet("Main St")
 *         .setCity("Springfield")
 *         .setState("IL")
 *         .setCountryCode("US")
 *         .setPostalCode("62704")
 *         .build();
 * </pre>
 * </p>
 */
public
class PostalAddressBuilder
{
    private String itsAddress;
    private String itsStreet;
    private String itsCity;
    private String itsState;
    private String itsCountryCode;
    private String itsPostalCode;

    public
    PostalAddressBuilder()
    {
        itsAddress     = null;
        itsStreet      = null;
        itsCity        = null;
        itsState       = null;
        itsCountryCode = null;
        itsPostalCode  = null;
    }

    public PostalAddressBuilder
    setAddress(String address)
    {
        itsAddress = address;
        return this;
    }

    public PostalAddressBuilder
    setStreet(String street)
    {
        itsStreet = street;
        return this;
    }

    public PostalAddressBuilder
    setCity(String city)
    {
        itsCity = city;
        return this;
    }

    public PostalAddressBuilder
    setState(String state)
    {
        itsState = state;
        return this;
    }

    public PostalAddressBuilder
    setCountryCode(String countryCode)
    {
        itsCountryCode = countryCode;
        return this;
    }

    public PostalAddressBuilder
    setPostalCode(String postalCode)
    {
        itsPostalCode = postalCode;
        return this;
    }

    public PostalAddress
    build()
    {
        return
            new PostalAddress(
                itsAddress,
                itsStreet,
                itsCity,
                itsState,
                itsCountryCode,
                itsPostalCode );
    }
}

//////////////////////////////////////////////////////////////////////////////

File: Components/Strata.Foundation.Core/main/typescript/Value/PostalAddressBuilder.ts
import {PostalAddress} from "./PostalAddress";

export
class PostalAddressBuilder
{
    private address: string;
    private street: string;
    private city: string;
    private state: string;
    private countryCode: string;
    private postalCode: string;

    constructor()
    {
        this.address = null;
        this.street = null;
        this.city = null;
        this.state = null;
        this.countryCode = null;
        this.postalCode = null;
    }

    setAddress(address: string): PostalAddressBuilder
    {
        this.address = address;
        return this;
    }

    setStreet(street: string): PostalAddressBuilder
    {
        this.street = street;
        return this;
    }

    setCity(city: string): PostalAddressBuilder
    {
        this.city = city;
        return this;
    }

    setState(state: string): PostalAddressBuilder
    {
        this.state = state;
        return this;
    }

    setCountryCode(countryCode: string): PostalAddressBuilder
    {
        this.countryCode = countryCode;
        return this;
    }

    setPostalCode(postalCode: string): PostalAddressBuilder
    {
        this.postalCode = postalCode;
        return this;
    }

    build(): PostalAddress
    {
        if (!this.address) throw Error("Address must be non-null");
        if (!this.street) throw Error("Street must be non-null");
        if (!this.city) throw Error("City must be non-null");
        if (!this.state) throw Error("State must be non-null");
        if (!this.countryCode) throw Error("CountryCode must be non-null");
        if (!this.postalCode) throw Error("PostalCode must be non-null");

        return {
            address: this.address,
            street: this.street,
            city: this.city,
            state: this.state,
            countryCode: this.countryCode,
            postalCode: this.postalCode
        };
    }
}